guides/c.md

```diff
@@ -1,10 +1,98 @@
-The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons.
+Introduction
+============
 
-- The projects we are wrapping usually have a style guide. Follow that, or make a best guess based on the code already written.
-- Usually the projects use clang-format. It's highly recommended to also use clang-format, and run it whenever the file is saved.
-- In the case of overrides, check if the override is a convenience override or not. If it is not, then ... (TODO: what guideline should be followed?)
+The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons. When in doubt, refer to to the wrapped library's style guides, or make a best guess based on code already written in either the wrapped library, or other wrappers.
 
-Here are some examples of how to style the C code.
+Formatting
+==========
+
+Use autoformatting where possible. Generally the wrapped libraries use clang-format, so it's best to use that. Also, it is suggested to do formatting on save.
+
+Implementation
+==============
+
+- If there are function overrides, then prefer to wrap the core implementation. If there's multiple implementations, then try to pick a suffix that is clear and understandable.
+
+Error Handling
+==============
+
+TODO
+
+SUGGESTION: Implement something like Rust's `Result<T, E>`, unless the wrapped library already implements error handling that we can inherit from in C.
```

Comment:
As attractive as this is, I don't think it's our job to "fix" libraries' error handling. This goes against your (very important) rule up above that the C layer should look as similar as possible to the underlying C++, and adding a different error-handling mechanism would be surprising for users of the API,  possibly very difficult to express in languages in other higher-level languages, and might even be impossible to represent in a Result-style.

Reply:
Catching exceptions and bundling them into ```Resullt<Value, std::string::String>``` is an option for the rust side right?

Reply:
That's true. Do any of the big libraries use exceptions as the primary error-handling mechanism?

Reply:
I remember seeing some exceptions in Imath. I could probably make the wording more clear, but the rule would be if the C++ can fail (through an exception), but has no error mechanism supported in C, then we have two options.

1. If we can use error mechanisms in the wrapped library to turn the C wrapper into a failable function, then do that.
2. If there are no clear ways to handle the error, then we need to implement our own design.

For example:

Error mechanism exists that we can use:

### C++

```cpp
bool can_fail();
void can_throw();

std::string get_error();
```

### C

```c
bool can_fail();
bool can_throw();

char* get_error();
```

Error methods don't exist at all:

### C++

```cpp
void can_throw();
```

### C

```c
enum StatusCode {
    OK;
    ERROR;
};
typedef struct {
    StatusCode status;
    union value {
        void value;
        Error err;
    }
} VoidStatus;

VoidStatus can_throw();
```

Reply:
I'm not clear about how we would generate the 'Error' type, in the last example for C.

Reply:
The error type could be a string or something else. I haven't thought too far ahead on what the error type is (if there will be one). But, the main takeaway is how do we handle errors if the wrapped library uses exceptions for error handling?

Reply:
The most common error-handling pattern in C is setting some internal error state and returning an error code, leaving it up to the caller to decide if they want to get the more detailed information or not. For a C++ function that might throw this would mean something like:

```
std::thread_local std::string error_string;

int some_function_that_throws() {
    try {
        CPP::some_function_that_throws();
    } catch (const Exception1& e) {
        error_string = e.what();
        return 1;
    } catch (const Exception2& e) {
        error_string = e.what();
        return 2;
    }

    return 0;
}

const char* get_error() {
      return error_string.c_str();
}
```

Reply:
Yes this makes more sense for c. Although the VoidStatus is cool in the sense that it's similar to how rust handles errors. Probably best to stick with what people are used to.

Reply:
Works for me. I'll update the document.